pkg/crd/flatten.go

```diff
@@ -439,3 +446,49 @@ func (f *flattenVisitor) Visit(baseSchema *apiext.JSONSchemaProps) SchemaVisitor
 
 	return f
 }
+
+// schemaOverride overrides the default schema for given a type and
+// flattening context. A non-nil value means override the default, else carry on
+// with the default flattening workflow for this type.
+// An example of an override will be to implement structuralschema for metadata
+// or embedded kubernetes types as defined in this KEP
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190425-structural-openapi.md
+func schemaOverride(ident TypeIdent, flattenCtx *flattenContext) *apiext.JSONSchemaProps {
+	metav1Pkg := "k8s.io/apimachinery/pkg/apis/meta/v1"
+	switch {
+	// root level ObjectMeta should not be specified in the validation schema
+	// because it is managed by kube-apiserver. Refer to the KEP specified
+	// above.
+	case flattenCtx.level == 1 /* root level metadata */ &&
+		ident.Package.ID == metav1Pkg && ident.Name == "ObjectMeta":
+		return &apiext.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]apiext.JSONSchemaProps{
+				"name": {
+					//TODO(droot): figure out a way to get the description
+					Type: "string",
+				},
+				"generateName": {
+					Type: "string",
+				},
+			},
+		}
+	case flattenCtx.level > 1 /* ObjectMeta being used at a non-root level */ &&
+		ident.Package.ID == metav1Pkg && ident.Name == "ObjectMeta":
+		return nil
+		// TODO(droot): enable it once we move k8s 1.15
+		// return &apiext.JSONSchemaProps{
+		// 	// XEmbeddedResource: true,
```

Comment:
for this @DirectXMan12 brought up the point of partial ObjectMeta. There is no requirement that the metadata is complete for an embedded resource. /cc @liggitt 

We either want a way to mark an embedded ObjectMeta as partial, or we relax the validation we do today (if that is possible now in beta). @liggitt wdyt?

Reply:
what is currently required that we would want to relax?

Reply:
as far as I remember, only kind/apiVersion were required

Reply:
It's about ObjectMeta: name, namespace as far as I see.

Reply:
name must be optional in openapi because of generateName, and iirc our server-side implementation put a dummy name in place when validating if no name was provided

namespace must already be optional (both server-side and in openapi) because of cluster-scoped resources

Reply:
ok, I think I misunderstood @DirectXMan12 in Slack. He was talking about `TypeMeta` (and I remember the dummy name value we put it to work around the check).

So the question is whether without TypeMeta it is really an embedded object.